dot/parachain/chainapi/chainapi.go

```diff
@@ -0,0 +1,133 @@
+
+type hashHeader struct {
+	Hash   common.Hash
+	Header types.Header
+}
+type ChainAPIMessage[message any] struct {
+	Message         message
+	ResponseChannel chan any
+}
+
+type AncestorsResponse struct {
+	Ancestors []common.Hash
+	Error     error
+}
+
+// Ancestors is a message to get the ancestors of a block.
+type Ancestors struct {
+	Hash              common.Hash
+	numberOfAncestors uint32
+}
+
+type BlockHeader common.Hash
+
+// DetermineNewBlocks determines the hashes of all new blocks we should track metadata for, given this head.
+//
+// This is guaranteed to be a subset of the (inclusive) ancestry of `head` determined as all
+// blocks above the lower bound or above the highest known block, whichever is higher.
+// This is formatted in descending order by block height.
+//
+// An implication of this is that if `head` itself is known or not above the lower bound,
+// then the returned list will be empty.
+//
+// This may be somewhat expensive when first recovering from major sync.
+func DetermineNewBlocks(subsystemToOverseer chan<- any, isKnown func(hash common.Hash) bool, head common.Hash,
+	header types.Header,
+	lowerBoundNumber parachaintypes.BlockNumber) ([]hashHeader, error) {
+	const maxNumberOfAncestors = 4
+	minBlockNeeded := uint(lowerBoundNumber + 1)
+
+	// Early exit if the block is in the DB or too early.
+	alreadyKnown := isKnown(head)
+	beforeRelevant := header.Number < minBlockNeeded
+	if alreadyKnown || beforeRelevant {
+		return nil, nil
+	}
+
+	ancestry := make([]hashHeader, 0)
+	headerClone, err := header.DeepCopy()
+	if err != nil {
+		return nil, fmt.Errorf("failed to deep copy header: %w", err)
+	}
+
+	ancestry = append(ancestry, hashHeader{Hash: head, Header: *headerClone})
+
+	// Early exit if the parent hash is in the DB or no further blocks are needed.
+	if isKnown(header.ParentHash) || header.Number == minBlockNeeded {
+		return ancestry, nil
+	}
+
+	lastHeader := ancestry[len(ancestry)-1].Header
+	// This is always non-zero as determined by the loop invariant above.
+	numberOfAncestors := min(maxNumberOfAncestors, (lastHeader.Number - minBlockNeeded))
+
+	ancestors, err := GetBlockAncestors(subsystemToOverseer, head, uint32(numberOfAncestors))
+	if err != nil {
+		return nil, fmt.Errorf("getting block ancestors: %w", err)
+	}
+	fmt.Printf("ancestors: %v\n", ancestors)
+	// TODO: finish this, see issue #3933
+
+	return ancestry, nil
+}
+
+// GetBlockAncestors sends a message to the overseer to get the ancestors of a block.
+func GetBlockAncestors(
+	overseerChannel chan<- any,
+	head common.Hash,
+	numAncestors uint32,
+) ([]common.Hash, error) {
+	respChan := make(chan any, 1)
+	message := ChainAPIMessage[Ancestors]{
+		Message: Ancestors{
+			Hash:              head,
+			numberOfAncestors: numAncestors,
+		},
+		ResponseChannel: respChan,
+	}
+	res, err := Call(overseerChannel, message, message.ResponseChannel)
+	if err != nil {
+		return nil, fmt.Errorf("sending message to get block ancestors: %w", err)
```

Comment:
when you call this function, you are wrapping an error with the `"getting block ancestors"` message. There is no need to wrap in this function. You can return the error without wrapping it.
```suggestion
		return nil, err
```

Reply:
updated.